### PR TITLE
[llvm-exegesis] Add option to specify the number of measurement repetitions

### PR DESCRIPTION
**Pull request by boomanaiden154**
Currently, the llvm-exegesis LatencyBenchmarkRunner repeats the benchmark several times (currently 30) and then aggregates the result to deal with noise in the measurement process. With this patch, the number of repetitions to perform is made configurable rather than left as a static number. This allows for significantly faster execution in situations where someone is performing a task like experimenting with memory annotations where the exact cycle counts might not be useful, and also allows for increased precision when desired.

### CODE REVIEW

File: llvm/tools/llvm-exegesis/lib/Target.h [hunk 1]
```diff
@@ -262,7 +262,8 @@ class ExegesisTarget {
       Benchmark::ModeE Mode, const LLVMState &State,
       BenchmarkPhaseSelectorE BenchmarkPhaseSelector,
       BenchmarkRunner::ExecutionModeE ExecutionMode,
-      Benchmark::ResultAggregationModeE ResultAggMode = Benchmark::Min) const;
+      Benchmark::ResultAggregationModeE ResultAggMode = Benchmark::Min,
+      unsigned BenchmarkRepeatCount = 30) const;
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/lib/Target.h:266]: Do we need the default value now that we have the flag in main ?

File: llvm/tools/llvm-exegesis/llvm-exegesis.cpp [hunk 2]
```diff
@@ -262,6 +262,12 @@ static cl::opt<BenchmarkRunner::ExecutionModeE> ExecutionMode(
                           "allows for the use of memory annotations")),
     cl::init(BenchmarkRunner::ExecutionModeE::InProcess));
 
+static cl::opt<unsigned>
+    BenchmarkRepeatCount("benchmark-repeat-count",
+                         cl::desc("The number of times to repeat the benchmark "
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/llvm-exegesis.cpp:267]: I would phrase this as "number of times to repeat measurements on the benchmark before aggregating the results", to make the difference with `-num-repetitions` clearer.

File: llvm/tools/llvm-exegesis/lib/Target.h (hunk 1 above)
Comment (boomanaiden154) [reply llvm/tools/llvm-exegesis/lib/Target.h]: Shouldn't be needed. I've removed it.

Also caught that I never plumbed the flag into the call in `llvm-exegesis.cpp`, so I've also added that in this patch.

File: llvm/tools/llvm-exegesis/llvm-exegesis.cpp (hunk 2 above)
Comment (boomanaiden154) [reply llvm/tools/llvm-exegesis/llvm-exegesis.cpp]: Good point. Fixed in the latest revision.

File: llvm/tools/llvm-exegesis/lib/LatencyBenchmarkRunner.h [hunk 3]
```diff
@@ -24,14 +24,16 @@ class LatencyBenchmarkRunner : public BenchmarkRunner {
...
+                         ExecutionModeE ExecutionMode,
+                         unsigned BenchmarkRepeatCount);
   ~LatencyBenchmarkRunner() override;
 
 private:
   Expected<std::vector<BenchmarkMeasure>>
   runMeasurements(const FunctionExecutor &Executor) const override;
 
   Benchmark::ResultAggregationModeE ResultAggMode;
+  unsigned BenchmarkRepetitions;
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/lib/LatencyBenchmarkRunner.h:36]: Similarly to the previous comment on the flag, let's call this `NumMeasurements` to differenciate with the repetitions within the snippet.

**Review by legrosbuffle — APPROVED** [review]

**Pull request merged**